dcp_prototype/backend/wrangling/migrations/scripts/copy_data_files.py

```diff
@@ -0,0 +1,154 @@
+            Bucket='dunitz-prod-copy',
+            Prefix=f"{project}/blobs",
+            ContinuationToken=continuation_token
+        )
+    else:
+        response = s3.list_objects_v2(
+            Bucket='dunitz-prod-copy',
+            Prefix=f"{project}/blobs"
+        )
+    if response['KeyCount'] > 0:
+        for x in response['Contents']:
+            keys.append(x['Key'])
+    if response['IsTruncated'] is True:
+        list_files_in_bucket(project, response['NextContinuationToken'], keys)
+    return keys
+
+
+def copy_between_s3_buckets(key: str, project: str, max_concurrency: int = 20):
+    """
+    Copy file between s3 buckets
+    :param key: File name
+    :param project: Project name
+    :param max_concurrency: Max concurrent threads for s3 bucket, boto3 default is 10
+    """
+    config = TransferConfig(max_concurrency=max_concurrency)
+
+    s3 = boto3.resource('s3')
+    copy_source = {
+        'Bucket': 'org-hca-dss-prod',
+        'Key': key
+    }
+    s3.meta.client.copy(copy_source, 'dunitz-prod-copy', f'{project}/{key}', Config=config)
+
+
+def list_data_files_for_project(directory_path: str) -> typing.List[typing.Dict[str, str]]:
+    """
+    Read all bundle manifests in a particular directory and return a list of non metadata files, file information
+    :param directory_path: Path to bundle manifest files
+    :return: List of dictionaries containing file information (hash, size, ids etc.) for all non metadata files for that project
+    """
+    project_data_files = []
+    for file in os.listdir(directory_path):
+        # skip hidden files
+        if file.startswith("."):
+            continue
+        file_path = os.path.join(directory_path, file)
+        with open(file_path, "rb") as file_reader:
+            contents = file_reader.read()
+            files = json.loads(contents)
+            for file_info in files['files']:
+                if "metadata" in file_info["content-type"]:
+                    continue
+                else:
+                    project_data_files.append(file_info)
+    return project_data_files
+
+
+def move_data_files(project_data_files: typing.List[str], project: str, precopied_files: typing.List[str]):
+    """
+    Parallelize calls to copy files between s3 buckets (if not already in destination bucket)
+    :param project_data_files: List of files to copy
+    :param project: Name of project (for destination key)
+    :param precopied_files: List of files already in destination bucket
+    """
+    files_moved = 0
+    already_there = 0
+    dispatch_executor_class = concurrent.futures.ThreadPoolExecutor
+    with dispatch_executor_class(max_workers=20) as executor:
+        print(f"{len(project_data_files)} files to copy over for {project}")
+        futures = []
+        for file_info in project_data_files:
+            blob_name = compose_blob_key(file_info)
+            key = f"{project}/{blob_name}"
+            if key in precopied_files:
```

Comment:
Question: how do you guarantee that the threads aren't copying over files that may be exactly the same? For example if thread_1 is copying over `project_0.json` from one bundle and thread_2 is copying over `project_0.json` from another bundle, the file contents will be the same but since neither copy has fully finished yet, wouldn't you end up with two copies of `project_0.json`?

Reply:
I think you are right. I refactored to create a set of blob_names to avoid the issue. Thanks!